# Machine Learning Model

In [1]:
# Import dependencies for random forest

from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler,OneHotEncoder
from sklearn.datasets import make_classification
import sklearn as skl
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import confusion_matrix, accuracy_score, classification_report


In [2]:
# dependencies for sql
import sqlalchemy as sql
from sqlalchemy import create_engine
import psycopg2
from sqlalchemy import MetaData

In [3]:
# other dependencies
import pandas as pd
import numpy as np
import csv
import os

In [4]:
# Import AWS password
from config import key

# Connecting database (AWS)

In [5]:
#Connection to the aws to import the data
engine = sql.create_engine('postgresql://ElectionTeam:finalproject@final-data-project.cxtxzq2kyq9u.us-east-1.rds.amazonaws.com:5432/election_data')

# Reading database

In [6]:
#Importing the database from AWS
elec_df=pd.read_sql_table("main_database", con= engine)
elec_df.head()

county_name  POPESTIMATE2019  INTERNATIONALMIG2019  DEMOCRAT  REPUBLICAN  \
0     AUTAUGA            55869                   -16    7503.0     19838.0   
1     BULLOCK            10101                    -1    3446.0      1146.0   
2     CHILTON            44428                    35    3073.0     16085.0   
3     COLBERT            55241                    73    8343.0     19203.0   
4     CONECUH            12067                     1    2966.0      3442.0   

   LIBERTARIAN  GREEN  OTHER  WINNER  Debt_All Debt_Communities_Color  \
0          0.0    0.0  429.0       0  .2623955                     49   
1          0.0    0.0   21.0       1  .2777778               .2727273   
2          0.0    0.0  152.0       0  .2815014                     49   
3          0.0    0.0  340.0       0  .2895472                     49   
4          0.0    0.0   33.0       0  .3046358               .3030303   

  Debt_Majorities_White  Avg_Household_Income_All  Avg_Non_Hispanic  \
0              .2649472                     75326             82971   
1                    49                     48571             80344   
2               .280027                     62784             65857   
3              .2890272                     60904             65431   
4                    .3                     47286             54421   

  Avg_Non_White_Income  
0                50809  
1                39215  
2                47591  
3                42998  
4                40122

In [7]:
# veryfing length
len(elec_df)

1779

# Preprocessing the data

Our goal is to predict if a ccounty were the Democrat Party won, won based on information we have in our DataFrame. To do this, we'll have to preprocessing the data, split our dataset into features (or inputs) and target (or outputs). 

The target, y, will be "winner", so we need to take out the winner columns. But also the parties columns because they are totally related with "winner", and we do not want to bias the predictive model. 

All the variables will help us determine if a county where the Democratic Party won has a population that is aligned with the party's ideology. In this sense, it is believed that the Democrats are more liberal and receptive of the immigrant population and minorities.

In [11]:
#Preprocessing the data, we change the values of 49 to zero so it could be in the same format as the other values, the values gos from 0 to 1, because they are rates.
elec_df1=elec_df.replace(49,0)
elec_df1.head()

county_name  POPESTIMATE2019  INTERNATIONALMIG2019  DEMOCRAT  REPUBLICAN  \
0     AUTAUGA            55869                   -16    7503.0     19838.0   
1     BULLOCK            10101                    -1    3446.0      1146.0   
2     CHILTON            44428                    35    3073.0     16085.0   
3     COLBERT            55241                    73    8343.0     19203.0   
4     CONECUH            12067                     1    2966.0      3442.0   

   LIBERTARIAN  GREEN  OTHER  WINNER  Debt_All Debt_Communities_Color  \
0          0.0    0.0  429.0       0  .2623955                     49   
1          0.0    0.0   21.0       1  .2777778               .2727273   
2          0.0    0.0  152.0       0  .2815014                     49   
3          0.0    0.0  340.0       0  .2895472                     49   
4          0.0    0.0   33.0       0  .3046358               .3030303   

  Debt_Majorities_White  Avg_Household_Income_All  Avg_Non_Hispanic  \
0              .2649472                     75326             82971   
1                    49                     48571             80344   
2               .280027                     62784             65857   
3              .2890272                     60904             65431   
4                    .3                     47286             54421   

  Avg_Non_White_Income  
0                50809  
1                39215  
2                47591  
3                42998  
4                40122

In [ ]:
#Drop the columns that won´t be used as features
election_df=elec_df.drop(columns=["county_name","DEMOCRAT","REPUBLICAN","LIBERTARIAN","GREEN","OTHER"])
election_df.head()

# Defining Tarjet and Features

The features set, X, are:

1. POPESTIMATE2019: population estimate for 2019.
2. INTERNATIONALMIG2019: international migration estimate for 2019 per each county.
3. Debt_All: rate of the county population with credit debt.
4. Debt_Communities_Color: county population rate disaggregated by communities of color with credit debt.
5. Debt_Majorities_White: county population rate disaggregated by communities with majorities white with credit debt.
6. Avg_Household_Income_All: county population average income.
7. Avg_Non_Hispanic: average income of non hispanic population.
8. Avg_Non_White_Income: average income of non white population.

The target set is the winner column, indicating whether or not a county won the Democrat party. If the Democrat Party lost the county (0) and if it won (1).

In [ ]:
#Define target and features
y=election_df["WINNER"]
X=election_df.drop(columns="WINNER")

In [ ]:
X.head()

In [ ]:
# We define the target set using the ravel() method, which performs the same procedure on our target set data as the values attribute
y=election_df["WINNER"].ravel()
y[:10]

# Random forest model

- We chose a Random forest model because of it´s high accuracy and interpretability. It can easily handle non linear data and outliers. The input will be in the form of tabular data (no images or natural language). A random forest model with a sufficient number of estimators and tree depth should be able to perform at a similar capacity to most deep learning models.
- We want a classifiter that predict which political party will win given certain conditions (features).
- We need to consider if the data classes have imbalance to apply an oversampling or undersampling technique. 
- The data will have over 1779 rows and 8 feature columns.
- After having everything in numerical values, we will standarized the data. 
- We will keep the default percentage of training and testing data (75% for training and 25% for testing)


## Split the data for train and test

To train and validate our model, we'll need to split the features and target sets into training and testing sets. This will help determine the relationships between each feature in the features training set and the target training set, which we'll use to determine the validity of our model using the features and target testing sets.

In [ ]:
#Use sklearn to split dataset for train and test
X_train, X_test, y_train, y_test = train_test_split(X,y, random_state=0)

- When the train_test_split() function is executed, our data is split into a specific proportion of the original data sets. By default, our training and testing data sets are 75% and 25%, respectively, of the original data.

In [ ]:
# Determine the shape of our training and testing sets.
print(X_train.shape)
print(X_test.shape)
print(y_train.shape)
print(y_test.shape)

In [ ]:
#Create scaler instance
X_scaler=skl.preprocessing.StandardScaler()

#Fit the scaler
X_scaler.fit(X_train)

#Scale the data
X_train_scaled=X_scaler.transform(X_train)
X_test_scaled=X_scaler.transform(X_test)

## 1st Trial
## Fitting the Random Forest Model

Before we fit the random forest model to our X_train_scaled and y_train training data, we'll create a random forest instance using the random forest classifier.

The n_estimators will allow us to set the number of trees that will be created by the algorithm. Generally, the higher number makes the predictions stronger and more stable, but can slow down the output because of the higher training time allocated. The best practice is to use between 64 and 128 random forests, though higher numbers are quite common despite the higher training time.

The fist trial we are going to use 80 estimators.

In [ ]:
#First trial with n_stimators=80
# Create a random forest classifier.
rf_model = RandomForestClassifier(n_estimators=80, random_state=78)

# Fitting the model
rf_model = rf_model.fit(X_train_scaled, y_train)

## Making Predictions Using the Testing Data

In [ ]:
# Evaluate the model
y_pred = rf_model.predict(X_test_scaled)
y_pred

## Model Evaluation
we analyze how well our random forest model classifies lusing the confusion_matrix.

In [ ]:
# Calculating the confusion matrix
cm = confusion_matrix(y_test,y_pred )

# Create a DataFrame from the confusion matrix.
cm_df = pd.DataFrame(
    cm, index=["Actual 0", "Actual  1"], columns=["Predicted 0", "Predicted 1"]
)

cm_df

The results show that:

- Out of 360 counties were the democrat party won, (Actual 1), 337 were predicted to be true (Predicted 0), which we call true positives.
- Out of 360 counties were the democrat party won, (Actual ), 23 were predicted to be false (Predicted 1), which are considered false negatives.
- Out of 85 counties were the democrat party won (Actual 1), 37 were predicted to be false (Predicted 0) and are considered false positives.
- Out of 85 counties were the democrat party won (Actual 1), 48 were predicted to be true (Predicted 1) and are considered true negatives.

TP+TN/Total=
Predicted good= 
Predicted Bad =
Actual Good=
Actual Bad=
Total observations =445

In [ ]:
print(f" Random forest predictive accuracy: {accuracy_score(y_test,y_pred):.3f}")
print(classification_report(y_test,y_pred))

In [ ]:
#Second trial with n_stimators=200
# Create a random forest classifier.
rf_model = RandomForestClassifier(n_estimators=200, random_state=78)

# Fitting the model
rf_model = rf_model.fit(X_train_scaled, y_train)

# Evaluate the model
y_pred = rf_model.predict(X_test_scaled)
print(f" Random forest predictive accuracy: {accuracy_score(y_test,y_pred):.3f}")
print(classification_report(y_test,y_pred))

In [ ]:
#Third trial with n_estimators=128
# Create a random forest classifier.
rf_model = RandomForestClassifier(n_estimators=128, random_state=78)

# Fitting the model
rf_model = rf_model.fit(X_train_scaled, y_train)

# Evaluate the model
y_pred = rf_model.predict(X_test_scaled)
print(f" Random forest predictive accuracy: {accuracy_score(y_test,y_pred):.3f}")
print(classification_report(y_test,y_pred))

In [ ]:
#Evaluate the model
# Calculating the confusion matrix.
cm = confusion_matrix(y_test, y_pred)

# Create a DataFrame from the confusion matrix.
cm_df = pd.DataFrame(
    cm, index=["Actual 0", "Actual 1"], columns=["Predicted 0", "Predicted 1"])

cm_df

#### Rank the importance of features

In [ ]:
# Calculate feature importance in the Random Forest model.
#importances = sorted(rf_model.feature_importances_,reverse=True)
importances = sorted(zip(rf_model.feature_importances_, X.columns), reverse=True)
# We can sort the features by their importance.
importances


## Logistic Regression (Alternative model)

In [ ]:
from sklearn.linear_model import LogisticRegression
election_df1=elec_df.copy()
election_df1.head()

In [ ]:
#Define training and testing sets
X2_train, X2_test, y2_train, y2_test = train_test_split(X,y, random_state=0, stratify=y)

In [ ]:
#Create scaler instance
X2_scaler=skl.preprocessing.StandardScaler()

#Fit the scaler
X2_scaler.fit(X2_train)

#Scale the data
X2_train_scaled=X2_scaler.transform(X2_train)
X2_test_scaled=X2_scaler.transform(X2_test)

In [ ]:
classifier = LogisticRegression(solver="lbfgs", random_state=0)
classifier.fit(X2_train_scaled,y2_train)
y_pred2= classifier.predict(X2_test_scaled)

In [ ]:
print(f" Logistic Regression accuracy: {accuracy_score(y2_test, y_pred2):.3f}")
print(classification_report(y2_test, y_pred2))


In [ ]:
#Evaluate the model
# Calculating the confusion matrix.
cm2 = confusion_matrix(y2_test, y_pred2)

# Create a DataFrame from the confusion matrix.
cm_df2 = pd.DataFrame(
    cm2, index=["Actual 0", "Actual 1"], columns=["Predicted 0", "Predicted 1"])

cm_df2